File: Linked List/python/intersectionPointOfTwoList.py
# Problem Statement: Given the heads of two singly linked-lists headA and headB, return the node at which the two lists intersect.
# If the two linked lists have no intersection at all, return null.

# Problem Link: https://www.codingninjas.com/studio/problems/-intersection-of-two-linked-lists_630457?utm_source=striver&utm_medium=website&utm_campaign=a_zcoursetuf

class Node:
    def __init__(self, data=0, next=None):
        self.data = data
        self.next = next

def get_length(head):
    count = 0
    while head:
        count += 1
        head = head.next
    return count

def find_intersection(first_head, second_head):
    # FIRST APPROACH - USING LENGTH VARIABLE
    # first = get_length(first_head)
    # second = get_length(second_head)
    # 
    # while first > second:
    #     first -= 1
    #     first_head = first_head.next
    #     
    # while second > first:
    #     second -= 1
    #     second_head = second_head.next
    #     
    # while first_head and second_head:
    #     if first_head == second_head:
    #         return first_head.data
    #     first_head = first_head.next
    #     second_head = second_head.next
    #     
    # return -1
    
    # SECOND APPROACH - USING DUMMY NODES
    dummy1 = first_head
    dummy2 = second_head
    
    while dummy1 != dummy2:
        dummy1 = second_head if dummy1 is None else dummy1.next
        dummy2 = first_head if dummy2 is None else dummy2.next
        
    if dummy1:
        return dummy1.data
        
    return -1

File: Linked List/cpp/intersectionPointOfTwoList.cpp
// Problem Statement: Given the heads of two singly linked-lists headA and headB, return the node at which the two lists intersect.
// If the two linked lists have no intersection at all, return null.

// Problem Link: https://www.codingninjas.com/studio/problems/-intersection-of-two-linked-lists_630457?utm_source=striver&utm_medium=website&utm_campaign=a_zcoursetuf

#include <iostream>
using namespace std;

class Node {
public:
    int data;
    Node* next;
    Node(int data) {
        this->data = data;
        this->next = NULL;
    }
};

int getLength(Node* head) {
    int count = 0;
    while (head != NULL) {
        count++;
        head = head->next;
    }
    return count;
}

int findIntersection(Node* firstHead, Node* secondHead) {
    // FIRST APPROACH - USING LENGTH VARIABLE
    /*
    int first = getLength(firstHead);
    int second = getLength(secondHead);
    
    while (first > second) {
        first--;
        firstHead = firstHead->next;
    }
    
    while (second > first) {
        second--;
        secondHead = secondHead->next;
    }
    
    while (firstHead != NULL && secondHead != NULL) {
        if (firstHead == secondHead) {
            return firstHead->data;
        }
        firstHead = firstHead->next;
        secondHead = secondHead->next;
    }
    return -1;
    */
    
    // SECOND APPROACH - USING DUMMY NODES
    Node* dummy1 = firstHead;
    Node* dummy2 = secondHead;

    while (dummy1 != dummy2) {
        dummy1 = dummy1 == NULL ? secondHead : dummy1->next;
        dummy2 = dummy2 == NULL ? firstHead : dummy2->next;
    }

    if (dummy1 != NULL) {
        return dummy1->data;
    }
    
    return -1;
}
